packages/flutter/lib/src/services/image_stream.dart

```diff
@@ -274,3 +288,163 @@ class OneFrameImageStreamCompleter extends ImageStreamCompleter {
     });
   }
 }
+
+/// Manages the decoding and scheduling of image frames.
+/// 
+/// This class deals with 2 types of frames:
+///  * image frames - image frames of an animated image.
+///  * app frames - frames that the flutter engine is drawing to the screen to
+///    show the app GUI.
+///
+/// For single frame images the stream will only complete once.
+///
+/// For animated images, this class eagerly decodes the next image frame,
+/// and notifies the listeners that a new frame is ready on the first app frame
+/// that is scheduled after the image frame duration has passed.
+///
+/// Scheduling new timers only from scheduled app frames, makes sure we pause
+/// the animation when the app is not visible (as new app frames will not be
+/// scheduled).
+///
+/// See the following timeline example:
+///
+/// | Time | Event                                      | Comment                   |
+/// |------|--------------------------------------------|---------------------------|
+/// | t1   | App frame scheduled (image frame A posted) |                           |
+/// | t2   | App frame scheduled                        |                           |
+/// | t3   | App frame scheduled                        |                           |
+/// | t4   | Image frame B decoded                      |                           |
+/// | t5   | App frame scheduled                        | t5 - t1 < frameB_duration | 
+/// | t6   | App frame scheduled (image frame B posted) | t6 - t1 > frameB_duration |
+///
+class MultiFrameImageStreamCompleter extends ImageStreamCompleter {
+  /// Creates a image stream completer.
+  ///
+  /// Immediately starts decoding the first image frame when the codec is ready.
+  ///
+  /// [codec] is a future for an initialized [ui.Codec] that will be used to
+  /// decode the image.
+  /// [scale] is the linear scale factor for drawing this frames of this image
+  /// at their intended size.
+  MultiFrameImageStreamCompleter({
+    @required Future<ui.Codec> codec,
+    @required double scale,
+    InformationCollector informationCollector
+  }) :
+    _informationCollector = informationCollector,
+    _scale = scale,
+    _framesEmitted = 0,
+    _timer = null
+  {
```

Comment:
nit: this open brace should be on the previous line

Reply:
I believe dartfmt may even get this right if you put a comma at the end of the last named argument